src/dll.py

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python
+# -*- coding: utf -8 -*-
+"""Implementation of a doubly linked list"""
+
+
+class Node(object):
+    def __init__(self, value):
+        """Create an instance of Node."""
+        self.value = value
+        self.next_node = None
+        self.previous_node = None
+
+
+class DoublyLL(object):
+    def __init__(self, iterable=None):
+        """Create an instance of DoublyLL.
+            The optional parameter <iterable> needs to be an iterable."""
+        self.head_node = None
+        self.tail_node = None
+        self.length_list = 0
+        if isinstance(iterable, list):
+            for i in iterable:
+                self.push(i)
+        elif iterable is not None:
+            self.push(iterable)
+
+    def push(self, val):
+        """Insert a new node to the head of a doubly linked list."""
+        new_node = Node(val)
+        if len(self) == 0:
+            self.tail_node = new_node
+            self.head_node = new_node
+        else:
+            self.head_node.previous_node = new_node
+            new_node.next_node = self.head_node
+            self.head_node = new_node
+        self.length_list += 1
+        return self
+
+    def __len__(self):
+        """Return the length of a doubly linked list."""
+        return self.length_list
+
+    def append(self, val):
+        """Insert a new node to the tail of a doubly linked list."""
+        new_node = Node(val)
+        if len(self) == 0:
+            self.tail_node = new_node
+            self.head_node = new_node
+        else:
+            self.tail_node.next_node = new_node
+            new_node.previous_node = self.tail_node
+            self.tail_node = new_node
+        self.length_list += 1
+        return self
+
+    def pop(self):
+        """Remove a node from the head, return removed node value."""
+        if self.length_list == 0:
+            raise IndexError("can't pop off an empty list")
+        popped_node = self.head_node
+        self.head_node = self.head_node.next_node
+        if self.head_node:
+            self.head_node.previous_node = None
+        self.length_list -= 1
+        return popped_node.value
+
+    def shift(self):
+        """Remove a node from the tail, return removed node value."""
+        if self.length_list == 0:
+            raise IndexError("can't shift an empty list")
+        shifted_node = self.tail_node
+        self.tail_node = self.tail_node.previous_node
+        if self.tail_node:
+            self.tail_node.next_node = None
+        self.length_list -= 1
+        return shifted_node.value
+
```

Comment:
Looking at shift/pop and append/push, you can see why Justin asked a question in Week 2 Thursday lecture about how we could DRY this code out. These pairs of functions are nearly identical except for (correct and well executed, nice job guys) small attribute name differences between them.

There is definitely a way of reducing the repetition of this code which I think Cris will get into in lecture next week, so be looking out for that. If you want a preview, read the documentation on the builtin Python functions `getattr` and `setattr`.